<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">

    <title>Pipeline Documentation</title>
    <meta name="author" content="Johnny Bravo">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <link rel="stylesheet" href="css/stylesheet.css">
    <link rel="stylesheet" href="javascript/highlight/styles/dracula.css">
    <script src="javascript/highlight/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
</head>

<body>
<div id='content'>
    <h1 id="versioning">Software, Data, and Versions</h1>

    <a href="index.html">Return to Table of Contents</a>

    <h2 id="software-versions">Software &amp; Versions</h2>
    <table>
        <thead>
            <tr>
                <th>Software</th>
                <th>Version</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>Awk</td>
                <td>1.3.4</td>
            </tr>
            <tr>
                <td>Aws Cli</td>
                <td>2.0.0</td>
            </tr>
            <tr>
                <td>Bcftools</td>
                <td>1.10.2</td>
            </tr>
            <tr>
                <td>Bgzip</td>
                <td>1.10.2</td>
            </tr>
            <tr>
                <td>BWA MEM</td>
                <td>0.7.14-r1188</td>
            </tr>
            <tr>
                <td>Cat</td>
                <td>8.3</td>
            </tr>
            <tr>
                <td>Expansion Hunter</td>
                <td>4.0.2</td>
            </tr>
            <tr>
                <td>FastQC</td>
                <td>0.11.9</td>
            </tr>
            <tr>
                <td>Grep</td>
                <td>3.4</td>
            </tr>
            <tr>
                <td>Gunzip</td>
                <td>1.10</td>
            </tr>
            <tr>
                <td>Htslib</td>
                <td>1.10.2-3</td>
            </tr>
            <tr>
                <td>Java</td>
                <td>11.0.9.1</td>
            </tr>
            <tr>
                <td>MultiQC</td>
                <td>1.9</td>
            </tr>
            <tr>
                <td>Nextflow</td>
                <td>20.0.7-10</td>
            </tr>
            <tr>
                <td>Panelcn.MOPS</td>
                <td>1.8.0</td>
            </tr>
            <tr>
                <td>Picard CollectHsMetrics</td>
                <td>2.23.8</td>
            </tr>
            <tr>
                <td>Picard CollectWgsMetrics</td>
                <td>2.23.8</td>
            </tr>
            <tr>
                <td>Python3</td>
                <td>3.8.5</td>
            </tr>
            <tr>
                <td>R</td>
                <td>3.6.3</td>
            </tr>
            <tr>
                <td>Samtools</td>
                <td>1.10</td>
            </tr>
            <tr>
                <td>Sed</td>
                <td>4.7</td>
            </tr>
            <tr>
                <td>SnpEff</td>
                <td>5.0c</td>
            </tr>
            <tr>
                <td>SnpSift</td>
                <td>5.0c</td>
            </tr>
            <tr>
                <td>Strelka2</td>
                <td>2.9.10</td>
            </tr>
            <tr>
                <td>Subsample</td>
                <td>0.0.6</td>
            </tr>
            <tr>
                <td>TrimmomaticPE</td>
                <td>0.39</td>
            </tr>
            <tr>
                <td>Ubuntu</td>
                <td>20.04</td>
            </tr>
            <tr>
                <td>Wgsim</td>
                <td>1.10</td>
            </tr>
        </tbody>
    </table>

    <h2 id="links-to-source">Links to Source Code</h2>
    <ul>
        <li><a href="https://github.com/samtools/bcftools">Bcftools</a></li>
        <li><a href="https://github.com/lh3/bwa">BWA MEM</a></li>
        <li><a href="https://github.com/Illumina/ExpansionHunter">Expansion Hunter</a></li>
        <li><a href="https://github.com/s-andrews/FastQC">FastQC</a></li>
        <li><a href="https://github.com/samtools/htslib">Htslib</a></li>
        <li><a href="https://github.com/ewels/MultiQC">MultiQC</a></li>
        <li><a href="https://github.com/nextflow-io/nextflow">Nextflow</a></li>
        <li><a href="https://github.com/bioinf-jku/panelcn.mops">Panelcn.MOPS</a></li>
        <li><a href="https://github.com/broadinstitute/picard">Picard</a></li>
        <li><a href="https://github.com/samtools/samtools">Samtools</a></li>
        <li><a href="https://github.com/pcingola/SnpEff">SnpEff</a></li>
        <li><a href="https://github.com/Illumina/strelka">Strelka2</a></li>
        <li><a href="https://github.com/paulgb/subsample">Subsample</a></li>
        <li><a href="https://github.com/timflutre/trimmomatic">TrimmomaticPE</a></li>
        <li><a href="https://github.com/lh3/wgsim">Wgsim</a></li>
    </ul>

    <h2 id="references-used">Reference Data Used</h2>
    <ul>
        <li><p>Reference Genome</p></li>
        <ul>
          <li><p>hg19</p></li>
        </ul>
        <li><p>Target and Bait for Exome Metrics<li><p>
        <ul>
          <li><p>xgen-exome-research-panel-v2-probes</p></li>
          <li><p>xgen-exome-research-panel-v2-targets</p></li>
        </ul>
        <li><p>Variant Catalog for Expansion Hunter</p></li>
        <ul>
          <li><p>Currently default packaged with program</p></li>
        </ul>
        <li><p>snpEff Database:</p></li>
        <ul>
          <li><p>hg19</p></li>
        </ul>
        <li><p>dbNSFP for SnpSift</p></li>
        <ul>
          <li><p>4.1a</p></li>
        </ul>
        <li><details><p><summary>dbNSFP Fields Used:</summary></p>
        <ul>
            <li><p>aapos</p></li>
            <li><p>Ensembl_geneid</p></li>
            <li><p>Ensembl_transcriptid</p></li>
            <li><p>Ensembl_proteinid</p></li>
            <li><p>Uniprot_acc</p></li>
            <li><p>Uniprot_entry</p></li>
            <li><p>HGVSc_ANNOVAR</p></li>
            <li><p>HGVSp_ANNOVAR</p></li>
            <li><p>HGVSc_snpEff</p></li>
            <li><p>HGVSp_snpEff</p></li>
            <li><p>HGVSc_VEP</p></li>
            <li><p>HGVSp_VEP</p></li>
            <li><p>APPRIS</p></li>
            <li><p>GENCODE_basic</p></li>
            <li><p>TSL String</p></li>
            <li><p>VEP_canonical</p></li>
            <li><p>cds_strand</p></li>
            <li><p>refcodon</p></li>
            <li><p>codonpos</p></li>
            <li><p>codon_degeneracy</p></li>
            <li><p>SIFT4G_score</p></li>
            <li><p>SIFT4G_pred</p></li>
            <li><p>REVEL_score</p></li>
            <li><p>CADD_phred</p></li>
            <li><p>1000Gp3_AF</p></li>
            <li><p>1000Gp3<em>AFR</em>AC</p></li>
            <li><p>1000Gp3<em>AFR</em>AF</p></li>
            <li><p>1000Gp3<em>EUR</em>AC</p></li>
            <li><p>1000Gp3<em>EUR</em>AF</p></li>
            <li><p>1000Gp3<em>AMR</em>AC</p></li>
            <li><p>1000Gp3<em>AMR</em>AF</p></li>
            <li><p>1000Gp3<em>EAS</em>AC</p></li>
            <li><p>1000Gp3<em>EAS</em>AF</p></li>
            <li><p>1000Gp3<em>SAS</em>AC</p></li>
            <li><p>1000Gp3<em>SAS</em>AF</p></li>
            <li><p>gnomAD<em>exomes</em>AC</p></li>
            <li><p>gnomAD<em>exomes</em>AN</p></li>
            <li><p>gnomAD<em>exomes</em>AF</p></li>
            <li><p>gnomAD<em>exomes</em>nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>AFR_AC</p></li>
            <li><p>gnomAD<em>exomes</em>AFR_AN</p></li>
            <li><p>gnomAD<em>exomes</em>AFR_AF</p></li>
            <li><p>gnomAD<em>exomes</em>AFR_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>AMR_AC</p></li>
            <li><p>gnomAD<em>exomes</em>AMR_AN</p></li>
            <li><p>gnomAD<em>exomes</em>AMR_AF</p></li>
            <li><p>gnomAD<em>exomes</em>AMR_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>ASJ_AC</p></li>
            <li><p>gnomAD<em>exomes</em>ASJ_AN</p></li>
            <li><p>gnomAD<em>exomes</em>ASJ_AF</p></li>
            <li><p>gnomAD<em>exomes</em>ASJ_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>EAS_AC</p></li>
            <li><p>gnomAD<em>exomes</em>EAS_AN</p></li>
            <li><p>gnomAD<em>exomes</em>EAS_AF</p></li>
            <li><p>gnomAD<em>exomes</em>EAS_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>FIN_AC</p></li>
            <li><p>gnomAD<em>exomes</em>FIN_AN</p></li>
            <li><p>gnomAD<em>exomes</em>FIN_AF</p></li>
            <li><p>gnomAD<em>exomes</em>FIN_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>NFE_AC</p></li>
            <li><p>gnomAD<em>exomes</em>NFE_AN</p></li>
            <li><p>gnomAD<em>exomes</em>NFE_AF</p></li>
            <li><p>gnomAD<em>exomes</em>NFE_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>SAS_AC</p></li>
            <li><p>gnomAD<em>exomes</em>SAS_AN</p></li>
            <li><p>gnomAD<em>exomes</em>SAS_AF</p></li>
            <li><p>gnomAD<em>exomes</em>SAS_nhomalt</p></li>
            <li><p>gnomAD<em>exomes</em>POPMAX_AC</p></li>
            <li><p>gnomAD<em>exomes</em>POPMAX_AN</p></li>
            <li><p>gnomAD<em>exomes</em>POPMAX_AF</p></li>
            <li><p>gnomAD<em>exomes</em>POPMAX_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>flag</p></li>
            <li><p>gnomAD<em>genomes</em>AC</p></li>
            <li><p>gnomAD<em>genomes</em>AN</p></li>
            <li><p>gnomAD<em>genomes</em>AF</p></li>
            <li><p>gnomAD<em>genomes</em>nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>AFR_AC</p></li>
            <li><p>gnomAD<em>genomes</em>AFR_AN</p></li>
            <li><p>gnomAD<em>genomes</em>AFR_AF</p></li>
            <li><p>gnomAD<em>genomes</em>AFR_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>AMR_AC</p></li>
            <li><p>gnomAD<em>genomes</em>AMR_AN</p></li>
            <li><p>gnomAD<em>genomes</em>AMR_AF</p></li>
            <li><p>gnomAD<em>genomes</em>AMR_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>ASJ_AC</p></li>
            <li><p>gnomAD<em>genomes</em>ASJ_AN</p></li>
            <li><p>gnomAD<em>genomes</em>ASJ_AF</p></li>
            <li><p>gnomAD<em>genomes</em>ASJ_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>EAS_AC</p></li>
            <li><p>gnomAD<em>genomes</em>EAS_AN</p></li>
            <li><p>gnomAD<em>genomes</em>EAS_AF</p></li>
            <li><p>gnomAD<em>genomes</em>EAS_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>FIN_AC</p></li>
            <li><p>gnomAD<em>genomes</em>FIN_AN</p></li>
            <li><p>gnomAD<em>genomes</em>FIN_AF</p></li>
            <li><p>gnomAD<em>genomes</em>FIN_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>NFE_AC</p></li>
            <li><p>gnomAD<em>genomes</em>NFE_AN</p></li>
            <li><p>gnomAD<em>genomes</em>NFE_AF</p></li>
            <li><p>gnomAD<em>genomes</em>NFE_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>AMI_AC</p></li>
            <li><p>gnomAD<em>genomes</em>AMI_AN</p></li>
            <li><p>gnomAD<em>genomes</em>AMI_AF</p></li>
            <li><p>gnomAD<em>genomes</em>AMI_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>SAS_AC</p></li>
            <li><p>gnomAD<em>genomes</em>SAS_AN</p></li>
            <li><p>gnomAD<em>genomes</em>SAS_AF</p></li>
            <li><p>gnomAD<em>genomes</em>SAS_nhomalt</p></li>
            <li><p>gnomAD<em>genomes</em>POPMAX_AC</p></li>
            <li><p>gnomAD<em>genomes</em>POPMAX_AN</p></li>
            <li><p>gnomAD<em>genomes</em>POPMAX_AF</p></li>
            <li><p>gnomAD<em>genomes</em>POPMAX_nhomalt</p></li>
            <li><p>clinvar_id</p></li>
            <li><p>clinvar_clnsig</p></li>
            <li><p>clinvar_trait</p></li>
            <li><p>clinvar_review</p></li>
            <li><p>clinvar_hgvs</p></li>
            <li><p>clinvar<em>var</em>source</p></li>
            <li><p>clinvar<em>MedGen</em>id</p></li>
            <li><p>clinvar<em>OMIM</em>id</p></li>
            <li><p>clinvar<em>Orphanet</em>id</p></li>
            <li><p>Interpro_domain</li>
        </ul>
        <br />
        </details></li>
    </ul>

    <h2 id="panels-used">Panels Used</h2>
    <blockquote>
        <p>Genome England Signed off Panels</p>
    </blockquote>
    <table>
        <thead>
            <tr>
                <th>Panel</th>
                <th>Version</th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <td>Intellectual Disability</td>
                <td>3.2</td>
            </tr>
            <tr>
                <td>Parkinsonism</td>
                <td>1.68</td>
            </tr>
            <tr>
                <td>Early Onset Dystonia</td>
                <td>1.85</td>
            </tr>
            <tr>
                <td>Childhood Onset Dystonia or Chorea or Related Movement Disorder</td>
                <td>1.58</td>
            </tr>
            <tr>
                <td>Hereditary Ataxia - Adult Onset</td>
                <td>2.13</td>
            </tr>
            <tr>
                <td>Hereditary Ataxia and Cerebellar Anomalies - Childhood Onset</td>
                <td>6.51</td>
            </tr>
            <tr>
                <td>Ataxia and Cerebellar Anomalies - Narrow Panel</td>
                <td>2.35</td>
            </tr>
            <tr>
                <td>Amyotrophic Lateral Sclerosis - Motor Neuron Disease</td>
                <td>1.29</td>
            </tr>
            <tr>
                <td>Neurodegenerative Disorders - Adult Onset</td>
                <td>2.38</td>
            </tr>
        </tbody>
    </table>
    <a href="#">To Top</a>
</div>
</body>
</html>