# Citations


## [Nextflow 21.10.6](https://pubmed.ncbi.nlm.nih.gov/28398311/)

> Di Tommaso P, Chatzou M, Floden EW, Barja PP, Palumbo E, Notredame C. Nextflow enables reproducible computational workflows. Nat Biotechnol. 2017 Apr 11;35(4):316-319. doi: 10.1038/nbt.3820. PubMed PMID: 28398311.

## Pipeline tools

* [FastQC 0.11.9](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

* [TrimmomaticPE 0.39](https://doi.org/10.1093/bioinformatics/btu170)
    > Anthony M. Bolger, Marc Lohse, Bjoern Usadel, Trimmomatic: a flexible trimmer for Illumina sequence data, Bioinformatics, Volume 30, Issue 15, 1 August 2014, Pages 2114–2120, https://doi.org/10.1093/bioinformatics/btu170

* [BWA MEM 0.7.17-r1188](https://doi.org/10.1093/bioinformatics/btp324)
    > Heng Li, Richard Durbin, Fast and accurate short read alignment with Burrows–Wheeler transform, Bioinformatics, Volume 25, Issue 14, 15 July 2009, Pages 1754–1760, https://doi.org/10.1093/bioinformatics/btp324

* [SAMtools 1.10](https://pubmed.ncbi.nlm.nih.gov/19505943/)
    > Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R; 1000 Genome Project Data Processing Subgroup. The Sequence Alignment/Map format and SAMtools. Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: 10.1093/bioinformatics/btp352. Epub 2009 Jun 8. PubMed PMID: 19505943; PubMed Central PMCID: PMC2723002.

* [picard-tools 2.23.8](http://broadinstitute.github.io/picard)

* [Strelka2 2.9.10](https://doi.org/10.1038/s41592-018-0051-x)
    > Kim, S., Scheffler, K., Halpern, A.L. et al. Strelka2: fast and accurate calling of germline and somatic variants. Nat Methods 15, 591–594 (2018). https://doi.org/10.1038/s41592-018-0051-x

* [snpEff 5.0c](https://doi.org/10.4161/fly.19695)
    > Pablo Cingolani, Adrian Platts, Le Lily Wang, Melissa Coon, Tung Nguyen, Luan Wang, Susan J. Land, Xiangyi Lu & Douglas M. Ruden (2012) A program for annotating and predicting the effects of single nucleotide polymorphisms, SnpEff, Fly, 6:2, 80-92, DOI: 10.4161/fly.19695

* [ExpansionHunter 4.0.2](https://doi.org/10.1093/bioinformatics/btz431)
    > Egor Dolzhenko, Viraj Deshpande, Felix Schlesinger, Peter Krusche, Roman Petrovski, Sai Chen, Dorothea Emig-Agius, Andrew Gross, Giuseppe Narzisi, Brett Bowman, Konrad Scheffler, Joke J F A van Vugt, Courtney French, Alba Sanchis-Juan, Kristina Ibáñez, Arianna Tucci, Bryan R Lajoie, Jan H Veldink, F Lucy Raymond, Ryan J Taft, David R Bentley, Michael A Eberle, ExpansionHunter: a sequence-graph-based tool to analyze variation in short tandem repeat regions, Bioinformatics, Volume 35, Issue 22, 15 November 2019, Pages 4754–4756, https://doi.org/10.1093/bioinformatics/btz431

* [ExpansionHunterDenovo 0.9.0](https://doi.org/10.1186/s13059-020-02017-z)
    > Dolzhenko, E., Bennett, M.F., Richmond, P.A. et al. ExpansionHunter Denovo: a computational method for locating known and novel repeat expansions in short-read sequencing data. Genome Biol 21, 102 (2020). https://doi.org/10.1186/s13059-020-02017-z

* [Manta 1.6.0](https://doi.org/10.1093/bioinformatics/btv710)
    > Xiaoyu Chen, Ole Schulz-Trieglaff, Richard Shaw, Bret Barnes, Felix Schlesinger, Morten Källberg, Anthony J. Cox, Semyon Kruglyak, Christopher T. Saunders, Manta: rapid detection of structural variants and indels for germline and cancer sequencing applications, Bioinformatics, Volume 32, Issue 8, 15 April 2016, Pages 1220–1222, https://doi.org/10.1093/bioinformatics/btv710

* [Gauchian 1.0.2](https://doi.org/10.1101/2021.11.12.21266253)
    > Marco Toffoli, Xiao Chen,  Fritz J Sedlazeck, Chiao-Yin Lee, Stephen Mullin, Abigail Higgins, Sofia Koletsi, Monica Emili Garcia-Segura, Esther Sammler, Sonja W. Scholz, Anthony HV Schapira,  Michael A. Eberle,  Christos Proukakis, Comprehensive analysis of GBA using a novel algorithm for Illumina whole-genome sequence data or targeted Nanopore sequencing, Medrxiv, preprint 2021, https://doi.org/10.1101/2021.11.12.21266253

* [Cyrius 1.1.1](https://doi.org/10.1038/s41397-020-00205-5)
    > Chen, X., Shen, F., Gonzaludo, N. et al. Cyrius: accurate CYP2D6 genotyping using whole-genome sequencing data. Pharmacogenomics J 21, 251–261 (2021). https://doi.org/10.1038/s41397-020-00205-5

* [ClassifyCNV 1.1.1](https://doi.org/10.1038/s41598-020-76425-3)
    > Gurbich, T.A., Ilinsky, V.V. ClassifyCNV: a tool for clinical annotation of copy-number variants. Sci Rep 10, 20375 (2020). https://doi.org/10.1038/s41598-020-76425-3

* [Bcftools 1.10.2](https://doi.org/10.1093/bioinformatics/btr509)
    > Li H. (2011). A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. Bioinformatics (Oxford, England), 27(21), 2987–2993. https://doi.org/10.1093/bioinformatics/btr509

* [MultiQC 1.9](https://pubmed.ncbi.nlm.nih.gov/27312411/)
    > Ewels P, Magnusson M, Lundin S, Käller M. MultiQC: summarize analysis results for multiple tools and samples in a single report. Bioinformatics. 2016 Oct 1;32(19):3047-8. doi: 10.1093/bioinformatics/btw354. Epub 2016 Jun 16. PubMed PMID: 27312411; PubMed Central PMCID: PMC5039924.

## Python packages

* [Python 3](https://www.python.org/)

* [pysam](https://github.com/pysam-developers/pysam)

## R packages

* [R](https://www.R-project.org/)
    > R Core Team (2017). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria.

* [cn.MOPS 1.36.0](https://doi.org/10.1093/nar/gks003)
    > Günter Klambauer, Karin Schwarzbauer, Andreas Mayr, Djork-Arné Clevert, Andreas Mitterecker, Ulrich Bodenhofer, Sepp Hochreiter, cn.MOPS: mixture of Poissons for discovering copy number variations in next-generation sequencing data with a low false discovery rate, Nucleic Acids Research, Volume 40, Issue 9, 1 May 2012, Page e69, https://doi.org/10.1093/nar/gks003

* [CopyNumberPlots](https://doi.org/10.18129/B9.bioc.CopyNumberPlots)
    > Gel B, Magallon M (2021). CopyNumberPlots: Create Copy-Number Plots using karyoploteR functionality. R package version 1.10.0, https://github.com/bernatgel/CopyNumberPlots.

## Software packaging/containerisation tools

* [Docker](https://dl.acm.org/doi/10.5555/2600239.2600241)

## Annotation Databases/ References

* [Genome Build hs37d5](ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence)
    > **Link is to download the reference**

* [CADD v1.6](https://cadd.gs.washington.edu/)
    > Rentzsch, P., Schubach, M., Shendure, J. et al. CADD-Splice—improving genome-wide variant effect prediction using deep learning-derived splice scores. Genome Med 13, 31 (2021). https://doi.org/10.1186/s13073-021-00835-9

* [REVEL 1.0](http://dx.doi.org/10.1016/j.ajhg.2016.08.016)
    > Ioannidis, N., Rothstein, J., Pejaver, V., Middha, S., McDonnell, S., Baheti, S., Musolf, A., Li, Q., Holzinger, E., Karyadi, D., Cannon-Albright, L., Teerlink, C., Stanford, J., Isaacs, W., Xu, J., Cooney, K., Lange, E., Schleutker, J., Carpten, J., Powell, I., Cussenot, O., Cancel-Tassin, G., Giles, G., MacInnis, R., Maier, C., Hsieh, C., Wiklund, F., Catalona, W., Foulkes, W., Mandal, D., Eeles, R., Kote-Jarai, Z., Bustamante, C., Schaid, D., Hastie, T., Ostrander, E., Bailey-Wilson, J., Radivojac, P., Thibodeau, S., Whittemore, A. and Sieh, W., 2016. REVEL: An Ensemble Method for Predicting the Pathogenicity of Rare Missense Variants. The American Journal of Human Genetics, 99(4), pp.877-885.

* [PMC](https://www.ncbi.nlm.nih.gov/pmc/)

* [OMIM](https://www.omim.org/)

* [gnomAD 2.1.1](https://doi.org/10.1038/s41586-020-2308-7)
    > Karczewski, K.J., Francioli, L.C., Tiao, G. et al. The mutational constraint spectrum quantified from variation in 141,456 humans. Nature 581, 434–443 (2020). https://doi.org/10.1038/s41586-020-2308-7

* [ClinVar GRcH37_2021-04-18](https://doi.org/10.1093/nar/gkx1153)
    > Melissa J Landrum, Jennifer M Lee, Mark Benson, Garth R Brown, Chen Chao, Shanmuga Chitipiralla, Baoshan Gu, Jennifer Hart, Douglas Hoffman, Wonhee Jang, Karen Karapetyan, Kenneth Katz, Chunlei Liu, Zenith Maddipatla, Adriana Malheiro, Kurt McDaniel, Michael Ovetsky, George Riley, George Zhou, J Bradley Holmes, Brandi L Kattman, Donna R Maglott, ClinVar: improving access to variant interpretations and supporting evidence, Nucleic Acids Research, Volume 46, Issue D1, 4 January 2018, Pages D1062–D1067, https://doi.org/10.1093/nar/gkx1153
